### ISSUE TITLE
Support Module Injection

### ISSUE THREAD

**Issue by iainbrighton** [VirtualEngine/Lability#106]
Sometimes it would be handy to be able to also inject modules into VMs. As an example, we might want to copying the latest version of Pester or Lability __into__ the virtual machines.

This would require extending the configuration data further. For example:
```powershell
@{
  NonNodeData = @{
    Lability = @{
      Module = @(
        ## Copy Pester v3.3.14 into all VMs. If not present locally, download from the PS Gallery
        @{ Name = 'Pester'; RequiredVersion = '3.3.14'; Provider = 'PSGallery'; }
        ## Copy Netscaler v1.2 into all VMs. If not present locally (or less than v1.2), download directly from Github
        @{ Name = 'Netscaler'; MinimumVersion = '1.2'; Provider = 'GitHub'; Owner = 'devblackops'; Repository = 'NetScaler'; }
      )
    }
  }
}
```
The key thing here is to be able to maintain multiple module (and DSC resource) versions - even on v4 hosts. This will probably require changing the behaviour to cache the module zip files in `%ProgramData%\Lability` rather than extracting them (like we currently do with DSC resource modules).

**Comment by kilasuit** [comment]
This would be great to include and additionally would be great to have the ability to provide a file path to a module as well.

Though this then begs the question should we extend to include scripts as well?

**Comment by iainbrighton** [comment]
OK @kilasuit @grayzu - there's a big update inbound. I'll push the branch when I've finished the tests and writing some notes. Here's what's changed:

* There's a new `-ModulePath` parameter added to `Set-LabHostDefault`
 * This is the path used to cache downloaded PowerShell modules/DSC resources
 * Only the defined `DSCResource` definitions are downloaded
 * When downloading DSC resources, versioned copies are stored in this location
 * This permits supporting different module versions in different configurations
* The `DSCResource` node is now __required__ as we need to know what resources to copy in!
 * If there is no `DSCResource` node, it will still copy all the locally installed DSC resources (existing behaviour)
 * A warning message is displayed that this will change in future
* A `NonNodeData\Lability\Module` section is now supported
 * This supports defining PowerShell modules to inject into VHD(X)s
 * This is essentially the same as the `DSCResource` node and you could put PowerShell modules in the `DSCResource` node
 * I've added it to be explicit
* Both DSC resources and PowerShell modules now support `Provider = 'FileSystem'`
 * These modules __are not__ stored in the cache - they are copied into the VHD(X) during VM creation
 * It supports pointing to folder containing a PowerShell module or to a .zip module

The "feature" that is removed with this branch is the ability to install DSC resources on the host. As we're now caching modules, they don't need to be installed. This functionality will probably move into the LabilityBootstrap module. If you wanted to compile configurations on the host you'd also need this - unless you think it should remain here and can provide a good argument! Is there anything in this that alarms you or you feel is still missing?

**Comment by kilasuit** [comment]
Sounds great though only one small thing

> If there is no DSCResource node, it will still copy all the locally installed DSC resources (existing behaviour)

I would see benefit in not copying over locally installed DSC Resources as there are a number on my machines that I use that I won't require in Lab Images so should in my opinion only copy these as a last resort if the person building the Lab doesn't specify the DSCResources to put into the Lab.

Also with conflicting DSC module versions - It would be helpful that the DSCResouces in the .psd1 were the "Authoritative" versions. Would make it easier on the config in the ps1 to just have basic import-DSCResource -modulename mod1,mod2,mod3,mod4 etc
 without worrying about versions (as you've done that in the psd1 - if you catch what I mean??

**Comment by iainbrighton** [comment]
> only copy these as a last resort if the person building the Lab doesn't specify the DSCResources to put into the Lab

Correct - the locally installed DSC resources are ___now only copied___ if there is no `DSCResource = @()` defined.

> Also with conflicting DSC module versions - It would be helpful that the DSCResouces in the .psd1 were the "Authoritative" versions. Would make it easier on the config in the ps1 to just have basic import-DSCResource -modulename mod1,mod2,mod3,mod4 etc
without worrying about versions (as you've done that in the psd1 - if you catch what I mean??

I think I understand what you're saying :stuck_out_tongue:. It will __only copy the versions defined__ in the `DSCResource` node (unless there is no definition!). For example, this will download the 2.9.0.0 version of the xNetworking module from the PowerShell Gallery:
```powershell
DSCResource = @(
    @{ Name = 'xNetworking'; Provider = 'PSGallery';  RequiredVersion = '2.9.0.0'; }
)
```
This will download the __latest__ version of the xNetworking module from the PowerShell Gallery:
```powershell
DSCResource = @(
    @{ Name = 'xNetworking'; Provider = 'PSGallery'; }
)
```
Does this answer the question or have I missed something?

**Comment by kilasuit** [comment]
Spot on now to get my hands on this latest update and rebuild some labs :-)

**Comment by grayzu** [comment]
Looks great, Iain! If the Module path is actually a cache location and not a location where the user puts the modules to be used by lability you might want to name the property ModuleCacheLocation or something instead.



MarkG
On Wed, Jul 13, 2016 at 7:07 PM, Ryan Yates < notifications@github.com [notifications@github.com] > wrote:
Spot on now to get my hands on this latest update and rebuild some labs :-)

**Comment by iainbrighton** [comment]
@grayzu In what way is a "Location" different to a "Path"? Would `ModuleCachePath` work better?

**Comment by grayzu** [comment]
Yeah, that is probably better. It just sounded a little funny to me but path is definitely more typical.


MarkG On Thu, Jul 14, 2016 at 4:28 AM, Iain Brighton < notifications@github.com [notifications@github.com] > wrote:
@grayzu [https://github.com/grayzu] In what way is a "Location" different to a "Path"? Would ModuleCachePath work better?

**Comment by iainbrighton** [comment]
@kilasuit @grayzu I've pushed my branch here https://github.com/iainbrighton/Lability/tree/Issue106. I still need to update some docs and comment etc. However, I think this is functionally complete (famous last words!).

**Comment by iainbrighton** [comment]
@kilasuit @grayzu This functionality has now been merged into the dev branch as I've been using it internally for a couple of weeks. Let me know how you get on :wink:!

**Comment by iainbrighton** [comment]
Closing as merged in 0.10.0 and I've not heard any noises suggesting it doesn't work..
